rgsync/Connectors/redis_connector.py

```diff
@@ -0,0 +1,177 @@
+                msg = "Exactly once property is not valid for Redis cluster"
+                raise Exception(msg) from None
+            # Get the entry corresponding to shard id in exactly once table
+            elif((self.exactlyOnceTableName is not None) and
+                    isinstance(self.connection, RedisConnection)):
+                shardId = 'shard-%s' % hashtag()
+                res = self.session.execute_command('HGET', self.exactlyOnceTableName, shardId)
+                if res is not None:
+                    self.exactlyOnceLastId = str(res)
+                else:
+                    self.shouldCompareId = False
+
+        except Exception as e:
+            self.session = None             #Next time we will connect to database
+            self.exactlyOnceLastId = None
+            self.shouldCompareId = True if self.exactlyOnceTableName is not None else False
+            msg = "Exception occur while getting shard id for exactly once. Exception : {}".format(e)
+            WriteBehindLog(msg)
+            raise Exception(msg) from None
+
+        pipe = self.session.pipeline()
+        try:
+            # data is list of dictionaries
+            # iterate over one by one and extract a dictionary and process it
+            lastStreamId = None
+            for d in data:
+                d_val = d['value']              # get value of key 'value' from dictionary
+
+                lastStreamId = d.pop('id', None)    # pop the stream id out of the record
+                if self.shouldCompareId and CompareIds(self.exactlyOnceLastId, lastStreamId) >= 0:
+                    WriteBehindLog('Skip {} as it was already written to the backend'.format(lastStreamId))
+                    continue
+
+                self.shouldCompareId = False
+
+                # check operation permission, what gets replicated
+                op = d_val.pop(OP_KEY, None)        # pop the operation key out of the record 
+                if op not in self.supportedOperations:
+                    msg = 'Got unknown operation'
+                    raise Exception(msg) from None
+
+                pk = d_val.pop(SIMPLE_HASH_BACKEND_PK)
+                newKey = '{}:{}'.format(self.new_prefix, pk)
+
+                if op != OPERATION_UPDATE_REPLICATE:
+                    # pipeline key to delete 
+                    pipe.delete(newKey)
+                else:
+                    # pipeline key and field-value mapping to set
+                    pipe.hset(newKey, mapping=d_val)
+            
+            pipe.execute()
+
+            # make entry for exactly once. In case of Redis cluster exception will be raised already
+            if((self.exactlyOnceTableName is not None) and
+                    isinstance(self.connection, RedisConnection)):
+                # error handling is not done as once entry is created update will return 0 
+                res = self.session.execute_command('HSET', self.exactlyOnceTableName, shardId, lastStreamId)
```

Comment:
This need to happened as part of the transaction pipe, otherwise the transaction might be done but this hset is not and we might execute the same update twice.

Reply:
I will update the code